pkg/phlaredb/block_querier.go

```diff
@@ -1999,22 +2009,32 @@ func (b *singleBlockQuerier) getUniqueLabelsSets(postings index.Postings, names
 				}
 				return nil, err
 			}
-			matchedLabels = append(matchedLabels, &typesv1.LabelPair{
-				Name:  name,
-				Value: value,
-			})
+
+			// Manually expand the pool's length and add this label to the end.
+			// The pool will always have enough capacity for all the labels.
+			matchedLabelsPool = matchedLabelsPool[:len(matchedLabelsPool)+1]
+			matchedLabelsPool[len(matchedLabelsPool)-1].Name = name
+			matchedLabelsPool[len(matchedLabelsPool)-1].Value = value
 		}
 
-		fp := matchedLabels.Hash()
+		fp := matchedLabelsPool.Hash()
 		_, ok := (*fingerprints)[fp]
 		if ok {
 			continue
 		}
 		(*fingerprints)[fp] = struct{}{}
 
-		labelsSets = append(labelsSets, &typesv1.Labels{
-			Labels: matchedLabels,
-		})
+		// Manually copy every element from the pool to a new slice.
+		labels := &typesv1.Labels{
+			Labels: make([]*typesv1.LabelPair, 0, len(matchedLabelsPool)),
+		}
+		for _, label := range matchedLabelsPool {
+			labels.Labels = append(labels.Labels, &typesv1.LabelPair{
+				Name:  label.Name,
+				Value: label.Value,
+			})
+		}
+		labelsSets = append(labelsSets, labels)
```

Comment:
```suggestion
		labelsSets = append(labelsSets, matchedLabelsPool.CloneVT())
```

Reply:
Nice one, fixed in https://github.com/grafana/pyroscope/pull/2838/commits/0f0779b9be17013495bed6b2a6d294d2e22e9290